dask/dataframe/io/io.py

```diff
@@ -750,5 +755,245 @@ def sorted_division_locations(seq, npartitions=None, chunksize=None):
+def from_map(
+    func,
+    *iterables,
+    args=None,
+    meta=None,
+    divisions=None,
+    label=None,
+    token=None,
+    enforce_metadata=True,
+    **kwargs,
+):
+    """Create a DataFrame collection from a custom function map
+
+    Parameters
+    ----------
+    func : callable
+        Function used to create each partition. If ``func`` satisfies the
+        ``DataFrameIOFunction`` protocol, column projection will be enabled.
+    *iterables : Iterable objects
+        Iterable objects to map to each output partition. all iterables must
+        be the same length. This length determines the number of partitions
+        in the output collection (only one element of each iterable will
+        be passed to ``func`` for each partition).
+    args : list or tuple, optional
+        Positional arguments to broadcast to each output partition. Note
+        that these arguments will always be passed to ``func`` after the
+        ``iterables`` positional arguments.
+    $META
+    divisions : tuple, str, optional
+        Partition boundaries along the index.
+        For tuple, see https://docs.dask.org/en/latest/dataframe-design.html#partitions
+        For string 'sorted' will compute the delayed values to find index
+        values.  Assumes that the indexes are mutually sorted.
+        If None, then won't use index information
+    label : str, optional
+        String to use as the function-name label in the output
+        collection-key names.
+    token : str, optional
+        String to use as the "token" in the output collection-key names.
+    enforce_metadata : bool, default True
+        Whether to enforce at runtime that the structure of the DataFrame
+        produced by ``func`` actually matches the structure of ``meta``.
+        This will rename and reorder columns for each partition,
+        and will raise an error if this doesn't work or types don't match.
+    **kwargs:
+        Key-word arguments to broadcast to each output partition. These
+        same arguments will be passed to ``func`` for every output partition.
+
+    Examples
+    --------
+    >>> import pandas as pd
+    >>> import dask.dataframe as dd
+    >>> func = lambda x, size=0: pd.Series([x] * size)
+    >>> inputs = ["A", "B"]
+    >>> dd.from_map(func, inputs, size=2).compute()
+    0    A
+    1    A
+    0    B
+    1    B
+    dtype: object
+
+    See Also
+    --------
+    dask.dataframe.from_delayed
+    dask.layers.DataFrameIOLayer
+    """
+
+    # Input validation
+    if not callable(func):
+        raise ValueError("`func` argument must be `callable`")
+    lengths = set()
+    for i, iterable in enumerate(iterables):
+        if not isinstance(iterable, Iterable):
+            raise ValueError(
+                f"All elements of `iterables` must be Iterable, got {type(iterable)}"
+            )
+        try:
+            lengths.add(len(iterable))
+        except AttributeError:
+            iterables[i] = list(iterable)
```

Comment:
What would prompt this to hit? Can an `Iterable` really not implement `len`?

Reply:
A generator (like `(i for i in range(50))`) is techincally an `Iterator` (it will pass the instance check above), but it won't support `len`.  Therefore, this check is mostly intended to convert generators to lists.